compiler/rustc_codegen_llvm/src/gotoc/cbmc/goto_program/symtab_transformer/gen_c_transformer/name_transformer.rs

```diff
@@ -0,0 +1,245 @@
+
+/// Struct for replacing names with valid C identifiers for --gen-c-runnable.
+pub struct NameTransformer {
+    new_symbol_table: SymbolTable,
+    mapped_names: FxHashMap<String, String>,
+    used_names: FxHashSet<String>,
+}
+
+impl NameTransformer {
+    /// Transform all identifiers in the symbol table to be valid C identifiers.
+    pub fn transform(original_symbol_table: &SymbolTable) -> SymbolTable {
+        let new_symbol_table = SymbolTable::new(original_symbol_table.machine_model().clone());
+        NameTransformer {
+            new_symbol_table,
+            mapped_names: FxHashMap::default(),
+            used_names: FxHashSet::default(),
+        }
+        .transform_symbol_table(original_symbol_table)
+    }
+
+    /// Converts an arbitrary identifier into a valid C identifier.
+    fn normalize_identifier(&mut self, orig_name: &str) -> String {
+        assert!(!orig_name.is_empty(), "Received empty identifier.");
+
+        // If name already encountered, return same result
+        match self.mapped_names.get(orig_name) {
+            Some(result) => return result.clone(),
+            None => (),
+        }
+
+        // Don't tranform the `tag-` prefix of identifiers
+        let (prefix, name) = if orig_name.starts_with("tag-") {
+            (&orig_name[..4], &orig_name[4..])
+        } else {
+            ("", orig_name)
+        };
+
+        // Separate function name from variable name for CBMC
+        let (name, suffix) = {
+            let mut parts = name.split("::1::");
+            let name = parts.next().unwrap();
+            let suffix = parts.next();
+            assert!(parts.next().is_none(), "Found multiple occurrences of '::1::' in identifier.");
+            (name, suffix)
+        };
+
+        // We separately call fix_name on the main part of the name
+        // and the suffix. This allows us to use the :: separator
+        // between the two parts, and also ensure that the
+        // base name of a variable stays as the suffix of the unique name.
+        fn fix_name(name: &str) -> String {
+            // Convert non-(alphanumeric + underscore) characters to underscore
+            let valid_chars =
+                name.replace(|ch: char| !(ch.is_alphanumeric() || ch == '_' || ch == '$'), "_");
+
+            // If the first character is a number, prefix with underscore
+            let new_name = match valid_chars.chars().next() {
+                Some(first) => {
+                    if first.is_numeric() {
+                        let mut name = "_".to_string();
+                        name.push_str(&valid_chars);
+                        name
+                    } else {
+                        valid_chars
+                    }
+                }
+                None => "".to_string(),
+            };
+
+            // Replace reserved names with alternatives
+            let mut illegal_names = [("case", "case_"), ("default", "default_")];
```

Comment:
I imagine this is not exhaustive?

Reply:
Really, we should just use the list of C reserved words

Reply:
Created issue.